File: homework/thinhbka/lesson-8-shortest_path_union_find/703. Kth Largest Element in a Stream.py
import heapq
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.heap = []

    def add(self, val: int) -> int:
        if len(self.heap) == 0:
            self.heap.append(val)
            return val 
        if self.heap[0] <= val:
            heapq.heappop(self.heap)
            heapq.heappush(self.heap,val)
        return self.heap[0]

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: homework/tai_nguyen/lesson11/703_kth_largest_element_in_a_stream.cpp
#include <queue>
#include <vector>

using namespace std;

class KthLargest {
public:
  priority_queue<int, vector<int>, greater<int>> min_heap;
  int size;

  KthLargest(int k, vector<int> &nums) {
    size = k;
    for (auto i : nums) {
      add(i);
    }
  }

  int add(int val) {
    min_heap.push(val);
    while (min_heap.size() > size) {
      min_heap.pop();
    }

    return min_heap.top();
  }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */